### PR TITLE
Improve file utilization when using a JetStream stream as a KV.

### PR DESCRIPTION
**Pull request by derekcollison**
Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/filestore.go [hunk 1]
```diff
@@ -216,8 +218,12 @@ const (
 	defaultStreamBlockSize = 16 * 1024 * 1024 // 16MB
 	// Default for workqueue or interest based.
 	defaultOtherBlockSize = 8 * 1024 * 1024 // 8MB
+	// Default for KV based
+	defaultKVBlockSize = 8 * 1024 * 1024 // 8MB
```
Comment (matthiashanel) [inline server/filestore.go:222]: Should this be used somewhere?

File: server/filestore.go [hunk 2]
```diff
@@ -791,7 +800,6 @@ func (mb *msgBlock) rebuildState() (*LostStreamData, error) {
 
 			mb.msgs++
 			mb.bytes += uint64(rl)
-			mb.rbytes += uint64(rl)
```
Comment (matthiashanel) [inline server/filestore.go:794]: This was essentially done by line 670?
mb.rbytes = uint64(len(buf))

File: server/filestore.go (hunk 1 above)
Comment (derekcollison) [reply server/filestore.go:222]: It is in stream.go

File: server/filestore.go (hunk 2 above)
Comment (derekcollison) [reply server/filestore.go:794]: Yes that was a bug, could not figure out why my compaction and rebuild state was saying same for rbytes ;)

**Review by matthiashanel — APPROVED** [review]
LGTM

File: server/filestore.go [hunk 3]
```diff
@@ -1870,6 +1889,93 @@ func (fs *fileStore) removeMsg(seq uint64, secure, needFSLock bool) (bool, error
...
+
+	// We will write to a new file and mv/rename it in case of failure.
+	mfn := path.Join(path.Join(mb.fs.fcfg.StoreDir, msgDir), fmt.Sprintf(newScan, mb.index))
+	defer os.Remove(mfn)
+	if err := ioutil.WriteFile(mfn, nbuf, defaultFilePerms); err != nil {
+		return
+	}
+	os.Rename(mfn, mb.mfn)
+
+	// Close cache and open FDs and index file.
```
Comment (kozlovic) [inline server/filestore.go:1966]: Should not closing be done before the rename? For instance, I don't think if this would work on Windows.

**Review by kozlovic — APPROVED** [review]
LGTM, but wonder if we should not be closing fds before the rewrite/rename, with possibly reopening in failed to rename, etc..

File: server/filestore.go (hunk 3 above)
Comment (derekcollison) [reply server/filestore.go:1966]: I will move it up.

**Review by ripienaar — APPROVED** [review]
LGTM

**Review by ripienaar — APPROVED** [review]
LGTM

**Pull request merged**